Adding scroll feature

Mohit-Agarwa1 opened the pull request:
Adding an 'offset' to the quotes draw function in screen.go in order to
allow moving down the list of tickers, using PgUp/PgDown or the Up/Down
arrow keys.

Currently, the offset is changed by 10 each time, and this is a hard
coded setting, but it would make sense to add this to the profile.

Also, scrolling with the mouse is not yet implemented, yet is possible
with termbox.

Initially, I set the up/down arrow to move one line at a time, but
moving one line and then trying to redraw the entire screen was
slow/laggy, so it moves 10 lines at a time currently. It might be nice
to have it move one line at a time, more line top/htop.

Mohit-Agarwa1 commented:
Fixed some of the issues mentioned in the pr in the last commit.

brandleesee commented:
Thank you @Mohit-Agarwa1 for tackling this enhancement request, may I ask whether you intend to make a setting in the profile for this and/or if you are also going to try and implement mouse-scrolling, please?
No pressure or commitment, I am only asking so I know how to proceed.

If yes to any of the above, how would you wish to go about them?
Would you rather keep adding to this PR or do you prefer to follow with a PR for each improvement?

Thank you once again and welcome on board.

Mohit-Agarwa1 commented:
Hi, in the commit 'Improved Scrolling' I've added the setting to profile and implemented mouse scrolling, so I think that this PR is done, and I'll open a new one if anything comes up or if I have any improvements.

Thanks

brandleesee commented:
Noted. I'll test it _with a _huge_ number of stocks in ticker_  and get back to you.
Thank you once again.

brandleesee commented:
![1](https://user-images.githubusercontent.com/12674437/166476975-c8a28624-f310-4b59-8db0-47b7d6150e88.gif)

There is a bug where on reaching the end of the list, there are residual characters.
I used the down and up arrow keys to test.

Can you look into this, @Mohit-Agarwa1, please?

Thank you very much.

brandleesee commented:
Please find results of my testing:

![3](https://user-images.githubusercontent.com/12674437/166803423-74782844-1fb5-416a-8237-818d555ec3b4.gif)
Scrolling on the same screen.

![4](https://user-images.githubusercontent.com/12674437/166804087-7ad5c216-a7cd-4a27-9af3-27f4954c0190.gif)
Scrolling after using on a secondary screen.

![5](https://user-images.githubusercontent.com/12674437/166806362-ec5e4642-0e1a-4563-a9e1-654181de8fb7.gif)
When the executable is on the secondary screen, then the window is minimised, and then maximised again, I observed automatic resizing of the text that fills the window.

Regarding the code optimisation mentioned in your last comment, please take your time.

Mohit-Agarwa1 commented:
Hi @brandleesee I've been trying to reproduce what you've been seeing with the two monitors. If I understand right, it's that there's a resolution difference between the monitors and when moving the window from one to the other the change in height is not recognised. I don't have two monitors of different resolutions, but every time I move the screen between monitors or slightly change the window size, the program runs Resize() and adjusts the line length.

It would be helpful if you could elaborate a little on what you're seeing, as I might not have understood what you've seen on your end. Otherwise, I would suggests that it's an interaction between your terminal and termbox.EventResize.

brandleesee commented:
Thank you @Mohit-Agarwa1 for the patches and your continued interest in mop. I am grateful. To elaborate more on my setup, the main display (laptop) is 1920x1080 however the other display is a bigger yet older one having a resolution of 1440x900.

I have tested your latest code (https://github.com/mop-tracker/mop/pull/100/commits/4512f396380a595a97983baa026c861db8ca3b86 https://github.com/mop-tracker/mop/pull/100/commits/83b716e9a4fb853eaf380658dea6a1485e71dd01) and observed the following anomalies. 
_This testing was carried out on the laptop screen only as at time of testing was not at my normal working station._

**Possible Bug**
The last line is hidden because of the taskbar.
To elaborate, the last symbol, 'V', is not visible. Apparently, this is because of the taskbar. However, when I AutoHide the taskbar, then 'V' is visible. This would imply that mop is being drawn until the bottom of the screen ignoring the taskbar.
![9](https://user-images.githubusercontent.com/12674437/167256530-3be15b87-0ea9-4bb0-a5c1-64fed2bc5614.gif)

 **Unable to resize window**
Moving, maximizing and resizing either produces garbled symbols or blanks out.
![10](https://user-images.githubusercontent.com/12674437/167256799-45774484-fb63-44da-9483-2cc0f419cdc7.gif)

_Other Observation_

From clicking to starting it seems slow.
I do not believe that this is due to a slow laptop (basic specs Asus 16GB i7-9750H / Windows 10).
I understand that this may not be due to your patches, so you may wish to ignore this observation.
![7](https://user-images.githubusercontent.com/12674437/167255760-bad07721-9d16-48df-943a-433a48b01b89.gif)

Please do not take my testing as negative criticism, on the contrary, this is all for having a polished piece of software and thank you once again for taking the time to work on mop.

Mohit-Agarwa1 commented:
Hi, thanks for the feedback again. Just thought it might be useful to show what I get from similar testing for the latest commit.


https://user-images.githubusercontent.com/46870579/167295155-8344bb2a-06dc-4c45-bf25-c91f028e5909.mp4

Mohit-Agarwa1 commented:
It seems that the video upload isn't working.
Regardless, I've changed the way screen resize works, so that when it resizes it doesn't try to fetch the markets, as when you change the screen size the program usually gets many resize events. In addition, I've made fetching the quotes concurrent so that you don't get those stutters when the queue updates.

brandleesee commented:
The transition when resizing the window is very smooth after these most recent changes, thank you.

The issue where the last stock is not visible because of the taskbar still persists.
(i.e. when I set the taskbar to autohide, the last stock 'V' becomes visible)
![1](https://user-images.githubusercontent.com/12674437/167301942-ee298461-d14e-4627-897f-ddf4f8d80377.gif)

brandleesee commented:
Partial success.
Scrolling with the down arrow, the last line is visible above the taskbar.
For some reason, scrolling down with the mouse results in the above gifs, ie the last line not showing.
Is there any fix for mouse scrolling, please?
Thank you very much.

brandleesee commented:
On closer inspection, it's the offset.

The arrow keys offset by 1 while mouse offsets by 5.

![image](https://user-images.githubusercontent.com/12674437/167914846-9ccadc78-282b-4a8d-8a55-fd718859086a.png)
main.go/arrow keys

![image](https://user-images.githubusercontent.com/12674437/167914913-b6f80d08-70ae-4a7c-b8b0-c069c553b16b.png)
main.go/mouse scroll

Furthermore, I re-tested that after scrolling (and the last ticker is again not displayed), just one press on the down arrow reveals it. Therefore, this would mean that the scroll offset of 5 limits scrolling to multiples of 5 not the remainder, example: for 16 lines it scrolls only until the 15th line.

Ideally, scrolling with the mouse should scroll until the bottom value regardless of the above calculation.

![2](https://user-images.githubusercontent.com/12674437/167915979-4d8d673e-dd49-4d94-b06c-7f120e803b72.gif)

brandleesee commented:
@Mohit-Agarwa1 may I ask whether my investigation above is enough to further patch this PR, please?
Thank you very much!

Mohit-Agarwa1 commented:
Hi, sorry for the inactivity. I think I know where the issue lies and
I'll get a patch in as soon as I can.

brandleesee commented:
Thank you for reaching back @Mohit-Agarwa1 and please take your time.

brandleesee commented:
@Mohit-Agarwa1 the below is what I get. Thank you very much!

Are there any other modifications you would like to do to this patch?

If not, may I apply the patch?

![1](https://user-images.githubusercontent.com/12674437/170659899-7e6a937a-810a-4d06-a1b0-f3ef59c0e714.gif)

Mohit-Agarwa1 commented:
No worries. I think the patch is good to go. Cheers.

brandleesee commented:
Merged, @Mohit-Agarwa1. Thank you very much for your support of this project :)
